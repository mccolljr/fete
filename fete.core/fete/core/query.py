"""Queries, the Q in CQRS."""

from typing import Any, Dict, Generic, Tuple, TypeVar, Union
from abc import ABCMeta, abstractmethod

from . import schema
from .context import Context

# pylint: disable=invalid-name
_T_Context = TypeVar("_T_Context", bound=Context)
_T_Result = TypeVar("_T_Result", bound=Union[schema.SchemaBase, None])
# pylint: enable=invalid-name


class QueryDefinitionError(Exception):
    """An exception caused by an invalid query class definition."""

    def __init__(self, query_name: str, problem: str):
        """Initialize a new QueryDefinitionError."""
        self.query_name = query_name
        self.problem = problem
        super().__init__(f"{self.query_name}: {self.problem}")


class QueryMeta(ABCMeta, schema.SchemaMeta):
    """The metaclass all query classes must have."""

    __query_mixin__: bool
    Result: schema.SchemaMeta

    def __new__(
        cls, name: str, bases: Tuple[type, ...], attrs: Dict[str, Any], **extra
    ):
        """Construct a new Query class."""
        attrs.setdefault("__query_mixin__", extra.pop("mixin", False))
        mewcls = super().__new__(cls, name, bases, attrs)
        return mewcls

    def __init__(
        cls, name: str, bases: Tuple[type, ...], attrs: Dict[str, Any], **_extra
    ):
        """Initialize and validate a newly created Query class."""
        super().__init__(name, bases, attrs)
        if not cls.__query_mixin__:
            QueryMeta._validate_fetch(cls)
            QueryMeta._validate_result(cls)

    @staticmethod
    def _validate_result(the_cls: type):

        result_class = getattr(the_cls, "Result", None)
        if result_class is None:
            raise QueryDefinitionError(the_cls.__name__, "Result must be defined")
        if not isinstance(result_class, schema.SchemaMeta):
            raise QueryDefinitionError(
                the_cls.__name__,
                f"Result must be a type with a schema (have {type(result_class)})",
            )

    @staticmethod
    def _validate_fetch(the_cls: type):
        fetch = getattr(the_cls, "fetch", None)
        if fetch is None or getattr(fetch, "__isabstractmethod__", False):
            raise QueryDefinitionError(the_cls.__name__, "fetch method must be defined")
        if not callable(fetch):
            raise QueryDefinitionError(the_cls.__name__, "fetch must be callable")


class QueryBase(Generic[_T_Context, _T_Result], schema.SchemaBase, metaclass=QueryMeta):
    """The class that all queries must subclass.

    Each subclass must specify a `Result` member which points to some class
    where `isinstance(Result, SchemaMeta)` is `True`

    If a subclass specifies a member `__query_mixin__` with a value of `True`,
    that class can serve as a common base class for multiple queries.
    """

    __query_mixin__ = True

    @abstractmethod
    async def fetch(self, context: _T_Context) -> _T_Result:
        """Implement the fetch operation for this query."""
        ...
