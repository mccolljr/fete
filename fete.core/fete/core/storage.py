"""The Storage protocol."""
from typing import Iterable, Optional, Protocol, TypeVar

from .predicate import Predicate


from .event import EventBase
from .aggregate import AggregateBase

# pylint: disable=invalid-name
_T_AggRoot = TypeVar("_T_AggRoot", bound=AggregateBase, covariant=True)
_T_EventRoot = TypeVar("_T_EventRoot", bound=EventBase, covariant=True)
# pylint: enable=invalid-name


class Storage(Protocol):
    """A protocol describing the shared capabilities of different storage providers."""

    async def load_events(
        self, query: Optional[Predicate] = None
    ) -> Iterable[_T_EventRoot]:
        """Load events that match the predicate."""
        ...

    async def save_events(self, events: Iterable[_T_EventRoot]):
        """Save new events."""
        ...

    async def save_snapshots(self, snaps: Iterable[_T_AggRoot]):
        """Save new snapshots."""
        ...

    async def load_snapshots(self, query: Predicate = None) -> Iterable[_T_AggRoot]:
        """Load snapshots that match the predicate."""
        ...
