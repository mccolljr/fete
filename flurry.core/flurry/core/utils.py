"""Utilities shared by storage solutions."""

# pylint: disable=invalid-name
from typing import (
    Any,
    Iterable,
    List,
    Optional,
    Protocol,
    Tuple,
    TypeVar,
    Union,
    cast,
)


from . import predicate as P

_T_co = TypeVar("_T_co", covariant=True)


class PredicateVisitor(Protocol[_T_co]):
    """The interface implemented by classes that visit the parts of a Predicate."""

    def on_is(self, p_is: P.Is) -> _T_co:
        """Process the predicate."""
        ...

    def on_or(self, p_or: P.Or) -> _T_co:
        """Process the predicate."""
        ...

    def on_and(self, p_and: P.And) -> _T_co:
        """Process the predicate."""
        ...

    def on_where(self, p_where: P.Where) -> _T_co:
        """Process the predicate."""
        ...

    def on_eq(self, field: str, p_eq: P.Eq) -> _T_co:
        """Process the field predicate."""
        ...

    def on_not_eq(self, field: str, p_neq: P.NotEq) -> _T_co:
        """Process the field predicate."""
        ...

    def on_less(self, field: str, p_less: P.Less) -> _T_co:
        """Process the field predicate."""
        ...

    def on_more(self, field: str, p_more: P.More) -> _T_co:
        """Process the field predicate."""
        ...

    def on_less_eq(self, field: str, p_less_eq: P.LessEq) -> _T_co:
        """Process the field predicate."""
        ...

    def on_more_eq(self, field: str, p_more_eq: P.MoreEq) -> _T_co:
        """Process the field predicate."""
        ...

    def on_between(self, field: str, p_between: P.Between) -> _T_co:
        """Process the field predicate."""
        ...

    def on_one_of(self, field: str, p_one_of: P.OneOf) -> _T_co:
        """Process the field predicate."""
        ...


def visit_predicate(visitor: PredicateVisitor[_T_co], pred: P.Predicate) -> _T_co:
    """Visit the predicate and its parts of the predicate using the given visitor."""
    if isinstance(pred, P.Is):
        return visitor.on_is(pred)
    if isinstance(pred, P.Or):
        return visitor.on_or(pred)
    if isinstance(pred, P.And):
        return visitor.on_and(pred)
    if isinstance(pred, P.Where):
        return visitor.on_where(pred)
    raise TypeError(f"unknown Predicate type {type(pred).__name__}")


def visit_field_predicate(
    visitor: PredicateVisitor[_T_co], field: str, pred: P.FieldPredicate
) -> _T_co:
    """Visit the field predicate using the given visitor."""
    if isinstance(pred, P.Eq):
        return visitor.on_eq(field, pred)
    if isinstance(pred, P.NotEq):
        return visitor.on_not_eq(field, pred)
    if isinstance(pred, P.Less):
        return visitor.on_less(field, pred)
    if isinstance(pred, P.More):
        return visitor.on_more(field, pred)
    if isinstance(pred, P.LessEq):
        return visitor.on_less_eq(field, pred)
    if isinstance(pred, P.MoreEq):
        return visitor.on_more_eq(field, pred)
    if isinstance(pred, P.Between):
        return visitor.on_between(field, pred)
    if isinstance(pred, P.OneOf):
        return visitor.on_one_of(field, pred)
    raise TypeError(f"unknown FieldPredicate type {type(pred).__name__}")


AnyPredicate = Union[P.Predicate, P.FieldPredicate]
SimplifiedPredicate = Tuple[
    Optional[AnyPredicate], Optional[str], Optional[Iterable[Any]]
]


def cast_simplified_predicate(src: SimplifiedPredicate):
    """Cast a SimplifiedPredicate value to one that explicitly deals with Predicate values."""
    pred, clause, params = src
    return cast(Optional[P.Predicate], pred), clause, params


def cast_simplified_field_predicate(src: SimplifiedPredicate):
    """Cast a SimplifiedPredicate value to one that explicitly deals with FieldPredicate values."""
    pred, clause, params = src
    return cast(Optional[P.FieldPredicate], pred), clause, params


class PredicateSQLSimplifier:
    """A base class for predicate visitors that perform SQL generation."""

    def on_is(self, p_is: P.Is) -> SimplifiedPredicate:
        """Do nothing."""
        return p_is, None, None

    def on_or(self, p_or: P.Or) -> SimplifiedPredicate:
        """Generate SQL representing the OR of all sub-predicates."""
        if not p_or.alts:
            return p_or, None, None
        items = [visit_predicate(self, p) for p in p_or.alts]
        preds: List[AnyPredicate] = []
        clauses: List[str] = []
        params: List[Any] = []
        for pr, c, pa in items:
            if pr is not None:
                preds.append(pr)
            if c is not None:
                clauses.append(c)
            if pa is not None:
                params.extend(pa)
        ret_pred = None
        ret_clauses = None
        ret_params = None
        if preds:
            ret_pred = P.Or(*cast(List[P.Predicate], preds))
        if clauses:
            ret_clauses = f"({' OR '.join(clauses)})"
        if params:
            ret_params = params
        return ret_pred, ret_clauses, ret_params

    def on_and(self, p_and: P.And) -> SimplifiedPredicate:
        """Generate SQL representing the AND of all sub-predicates."""
        if not p_and.preds:
            return p_and, None, None
        items = [visit_predicate(self, p) for p in p_and.preds]
        preds: List[AnyPredicate] = []
        clauses: List[str] = []
        params: List[Any] = []
        for pr, c, pa in items:
            if pr is not None:
                preds.append(pr)
            if c is not None:
                clauses.append(c)
            if pa is not None:
                params.extend(pa)
        ret_pred = None
        ret_clauses = None
        ret_params = None
        if preds:
            ret_pred = P.And(*cast(List[P.Predicate], preds))
        if clauses:
            ret_clauses = f"({' AND '.join(clauses)})"
        if params:
            ret_params = params
        return ret_pred, ret_clauses, ret_params

    def on_where(self, p_where: P.Where) -> SimplifiedPredicate:
        """Do nothing."""
        return p_where, None, None

    def on_eq(self, _field: str, p_eq: P.Eq) -> SimplifiedPredicate:
        """Do nothing."""
        return p_eq, None, None

    def on_not_eq(self, _field: str, p_neq: P.NotEq) -> SimplifiedPredicate:
        """Do nothing."""
        return p_neq, None, None

    def on_less(self, _field: str, p_less: P.Less) -> SimplifiedPredicate:
        """Do nothing."""
        return p_less, None, None

    def on_more(self, _field: str, p_more: P.More) -> SimplifiedPredicate:
        """Do nothing."""
        return p_more, None, None

    def on_less_eq(self, _field: str, p_less_eq: P.LessEq) -> SimplifiedPredicate:
        """Do nothing."""
        return p_less_eq, None, None

    def on_more_eq(self, _field: str, p_more_eq: P.MoreEq) -> SimplifiedPredicate:
        """Do nothing."""
        return p_more_eq, None, None

    def on_between(self, _field: str, p_between: P.Between) -> SimplifiedPredicate:
        """Do nothing."""
        return p_between, None, None

    def on_one_of(self, _field: str, p_one_of: P.OneOf) -> SimplifiedPredicate:
        """Do nothing."""
        return p_one_of, None, None
