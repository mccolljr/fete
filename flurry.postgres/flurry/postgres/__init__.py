from .postgres import _PostgreSQLSimplifier
