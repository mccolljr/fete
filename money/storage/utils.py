# pylint: disable=invalid-name
from __future__ import annotations
from typing import (
    Any,
    Iterable,
    List,
    Optional,
    Protocol,
    Tuple,
    TypeVar,
    Union,
    cast,
)


from money import predicate as P
from money.predicate import FieldPredicate, Predicate

T = TypeVar("T", covariant=True)


class PredicateVisitor(Protocol[T]):
    """
    A class with methods that can be called to process
    individual predicates while walking a predicate tree.
    """

    def on_is(self, p_is: P.Is) -> T:
        ...

    def on_or(self, p_or: P.Or) -> T:
        ...

    def on_and(self, p_and: P.And) -> T:
        ...

    def on_where(self, p_where: P.Where) -> T:
        ...

    def on_eq(self, field: str, p_eq: P.Eq) -> T:
        ...

    def on_not_eq(self, field: str, p_neq: P.NotEq) -> T:
        ...

    def on_less(self, field: str, p_less: P.Less) -> T:
        ...

    def on_more(self, field: str, p_more: P.More) -> T:
        ...

    def on_less_eq(self, field: str, p_less_eq: P.LessEq) -> T:
        ...

    def on_more_eq(self, field: str, p_more_eq: P.MoreEq) -> T:
        ...

    def on_between(self, field: str, p_between: P.Between) -> T:
        ...

    def on_one_of(self, field: str, p_one_of: P.OneOf) -> T:
        ...


def visit_predicate(visitor: PredicateVisitor[T], pred: Predicate) -> T:
    if isinstance(pred, P.Is):
        return visitor.on_is(pred)
    if isinstance(pred, P.Or):
        return visitor.on_or(pred)
    if isinstance(pred, P.And):
        return visitor.on_and(pred)
    if isinstance(pred, P.Where):
        return visitor.on_where(pred)
    raise TypeError(f"unknown Predicate type {type(pred).__name__}")


def visit_field_predicate(
    visitor: PredicateVisitor[T], field: str, pred: FieldPredicate
) -> T:
    if isinstance(pred, P.Eq):
        return visitor.on_eq(field, pred)
    if isinstance(pred, P.NotEq):
        return visitor.on_not_eq(field, pred)
    if isinstance(pred, P.Less):
        return visitor.on_less(field, pred)
    if isinstance(pred, P.More):
        return visitor.on_more(field, pred)
    if isinstance(pred, P.LessEq):
        return visitor.on_less_eq(field, pred)
    if isinstance(pred, P.MoreEq):
        return visitor.on_more_eq(field, pred)
    if isinstance(pred, P.Between):
        return visitor.on_between(field, pred)
    if isinstance(pred, P.OneOf):
        return visitor.on_one_of(field, pred)
    raise TypeError(f"unknown FieldPredicate type {type(pred).__name__}")


AnyPredicate = Union[Predicate, FieldPredicate]
SimplifiedPredicate = Tuple[
    Optional[AnyPredicate], Optional[str], Optional[Iterable[Any]]
]


def cast_simplified_predicate(src: SimplifiedPredicate):
    """Casts a SimplifiedPredicate value to one that explicitly deals with Predicate values."""
    pred, clause, params = src
    return cast(Optional[Predicate], pred), clause, params


def cast_simplified_field_predicate(src: SimplifiedPredicate):
    """Casts a SimplifiedPredicate value to one that explicitly deals with FieldPredicate values."""
    pred, clause, params = src
    return cast(Optional[FieldPredicate], pred), clause, params


class PredicateSQLSimplifier:
    """A base class for predicate visitors that perform SQL generation."""

    def on_is(self, p_is: P.Is) -> SimplifiedPredicate:
        return p_is, None, None

    def on_or(self, p_or: P.Or) -> SimplifiedPredicate:
        items = [visit_predicate(self, p) for p in p_or.alts]
        preds: List[AnyPredicate] = []
        clauses: List[str] = []
        params: List[Any] = []
        for pr, c, pa in items:
            if pr is not None:
                preds.append(pr)
            if c is not None:
                clauses.append(c)
            if pa is not None:
                params.extend(pa)
        ret_pred = None
        ret_clauses = None
        ret_params = None
        if preds:
            ret_pred = P.Or(*cast(List[Predicate], preds))
        if clauses:
            ret_clauses = f"({' OR '.join(clauses)})"
        if params:
            ret_params = params
        return ret_pred, ret_clauses, ret_params

    def on_and(self, p_and: P.And) -> SimplifiedPredicate:
        items = [visit_predicate(self, p) for p in p_and.preds]
        preds: List[AnyPredicate] = []
        clauses: List[str] = []
        params: List[Any] = []
        for pr, c, pa in items:
            if pr is not None:
                preds.append(pr)
            if c is not None:
                clauses.append(c)
            if pa is not None:
                params.extend(pa)
        ret_pred = None
        ret_clauses = None
        ret_params = None
        if preds:
            ret_pred = P.And(*cast(List[Predicate], preds))
        if clauses:
            ret_clauses = f"({' AND '.join(clauses)})"
        if params:
            ret_params = params
        return ret_pred, ret_clauses, ret_params

    def on_where(self, p_where: P.Where) -> SimplifiedPredicate:
        return p_where, None, None

    def on_eq(self, _field: str, p_eq: P.Eq) -> SimplifiedPredicate:
        return p_eq, None, None

    def on_not_eq(self, _field: str, p_neq: P.NotEq) -> SimplifiedPredicate:
        return p_neq, None, None

    def on_less(self, _field: str, p_less: P.Less) -> SimplifiedPredicate:
        return p_less, None, None

    def on_more(self, _field: str, p_more: P.More) -> SimplifiedPredicate:
        return p_more, None, None

    def on_less_eq(self, _field: str, p_less_eq: P.LessEq) -> SimplifiedPredicate:
        return p_less_eq, None, None

    def on_more_eq(self, _field: str, p_more_eq: P.MoreEq) -> SimplifiedPredicate:
        return p_more_eq, None, None

    def on_between(self, _field: str, p_between: P.Between) -> SimplifiedPredicate:
        return p_between, None, None

    def on_one_of(self, _field: str, p_one_of: P.OneOf) -> SimplifiedPredicate:
        return p_one_of, None, None
