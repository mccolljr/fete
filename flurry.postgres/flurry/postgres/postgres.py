"""PostgreSQL storage solution."""

from typing import Any, AsyncGenerator, Iterable, List, Optional, Tuple

import aiopg  # type: ignore
import logging
import datetime as dt
from contextlib import asynccontextmanager

from flurry.util import JSON, RWLock
from flurry.core import predicate as P
from flurry.core.event import EventBase, EventMeta
from flurry.core.predicate import Predicate
from flurry.core.aggregate import AggregateBase, AggregateMeta
from flurry.core.utils import (
    PredicateSQLSimplifier,
    SimplifiedPredicate,
    cast_simplified_predicate,
    visit_field_predicate,
    visit_predicate,
)

LOG = logging.getLogger("flurry.postgres")


class _PostgreSQLSimplifier(PredicateSQLSimplifier):
    def __init__(self, type_field: str, data_field: str, timestamp_convert: str = None):
        self.type_field = type_field
        self.data_field = data_field
        self.timestamp_convert = (
            timestamp_convert
            if timestamp_convert is not None
            else r"""to_timestamp({}, 'YYYY-MM-DD"T"HH24:MI:SSTZH:TZM')"""
        )

    def _ph(self) -> str:
        return "%s"

    def on_is(self, p_is: P.Is) -> SimplifiedPredicate:
        if not p_is.types:
            return p_is, None, None
        clause = f"{self.type_field} IN ({', '.join(self._ph() for _ in p_is.types)})"
        params = [t.__name__ if isinstance(t, type) else str(t) for t in p_is.types]
        return None, clause, params

    def on_where(self, p_where: P.Where) -> SimplifiedPredicate:
        if not p_where.fields:
            return p_where, None, None
        exprs: List[str] = []
        params: List[Any] = []
        for name, fpred in p_where.fields.items():
            pred, clause, fparams = visit_field_predicate(self, name, fpred)
            assert pred is None and clause is not None and fparams is not None
            exprs.append(clause)
            params.extend(fparams)
        return None, f"({' AND '.join(exprs)})", params

    def _smart_query(
        self, field: str, oper: str, val: Any
    ) -> Tuple[str, Iterable[Any]]:
        is_negation = oper in ("<>", "!=")
        if isinstance(val, (str, int, float, bool)):
            if is_negation:
                return (
                    f"coalesce({self.data_field}->{self._ph()} {oper} {self._ph()}::jsonb, true)",
                    (field, JSON.dumps(val)),
                )
            return (
                f"({self.data_field} ? {self._ph()} AND {self.data_field}->{self._ph()} {oper} {self._ph()}::jsonb)",
                (field, field, JSON.dumps(val)),
            )
        if isinstance(val, dt.datetime):
            field_as_timestamp_tz = self.timestamp_convert.format(
                f"{self.data_field}->>{self._ph()}"
            )
            if is_negation:
                return (
                    f"coalesce({field_as_timestamp_tz} {oper} {self._ph()}::timestamp, true)",
                    (field, val.isoformat()),
                )
            return (
                f"({self.data_field} ? {self._ph()} AND {field_as_timestamp_tz} {oper} {self._ph()}::timestamp)",
                (field, field, val.isoformat()),
            )
        raise RuntimeError(f"unsupported value type {type(val).__name__}")

    def on_eq(self, field: str, p_eq: P.Eq) -> SimplifiedPredicate:
        query, params = self._smart_query(field, "=", p_eq.expect)
        return (None, query, params)

    def on_not_eq(self, field: str, p_neq: P.NotEq) -> SimplifiedPredicate:
        query, params = self._smart_query(field, "<>", p_neq.value)
        return (None, query, params)

    def on_less(self, field: str, p_less: P.Less) -> SimplifiedPredicate:
        query, params = self._smart_query(field, "<", p_less.limit)
        return (None, query, params)

    def on_more(self, field: str, p_more: P.More) -> SimplifiedPredicate:
        query, params = self._smart_query(field, ">", p_more.limit)
        return (None, query, params)

    def on_less_eq(self, field: str, p_less_eq: P.LessEq) -> SimplifiedPredicate:
        query, params = self._smart_query(field, "<=", p_less_eq.limit)
        return (None, query, params)

    def on_more_eq(self, field: str, p_more_eq: P.MoreEq) -> SimplifiedPredicate:
        query, params = self._smart_query(field, ">=", p_more_eq.limit)
        return (None, query, params)

    def on_between(self, field: str, p_between: P.Between) -> SimplifiedPredicate:
        low_query, low_params = self._smart_query(field, ">=", p_between.lower)
        hi_query, hi_params = self._smart_query(field, "<=", p_between.upper)
        return (
            None,
            f"({low_query} AND {hi_query})",
            tuple(list(low_params) + list(hi_params)),
        )

    def on_one_of(self, field: str, p_one_of: P.OneOf) -> SimplifiedPredicate:
        if not p_one_of.options:
            return p_one_of, None, None
        params: List[Any] = []
        exprs: List[str] = []
        for opt in p_one_of.options:
            oquery, oparams = self._smart_query(field, "=", opt)
            exprs.append(oquery)
            params.extend(oparams)
        return (None, f"({' OR '.join(exprs)})", tuple(params))


class PostgreSQLStorage:
    """Provides a storage interface for a postgresql database."""

    def __init__(
        self, *, host: str, port: str, user: str, password: str, database: str, **pgopts
    ):
        """Initialize new PostgreSQL storage using the given connection."""
        self.__setup = RWLock()
        self.__dsn = (
            f"postgres://{user}:{password}@{host}:{port}/{database}"
            f"?{'&'.join(f'{name}={value}' for name, value in pgopts.items())}"
        )
        self.__pool: Optional[aiopg.Pool] = None
        self.__timestamp_convert: Optional[str] = None

    async def __get_pool(self) -> aiopg.Pool:
        async with self.__setup.read:
            if self.__pool is not None:
                return self.__pool
        async with self.__setup.write:
            if self.__pool is not None:
                return self.__pool
            pool: aiopg.Pool = await aiopg.create_pool(self.__dsn)
            async with pool.acquire() as conn:
                async with conn.cursor() as cur:
                    await self.__init_db(cur)
            self.__pool = pool
            return pool

    async def __init_db(self, cur: aiopg.Cursor):
        await cur.execute("BEGIN;")
        try:
            await cur.execute(
                """
                CREATE TABLE IF NOT EXISTS __events (
                    sequence_num BIGSERIAL NOT NULL PRIMARY KEY,
                    event_type   VARCHAR(128) NOT NULL,
                    event_data   JSONB NOT NULL DEFAULT '{}'
                );
                """
            )
            await cur.execute(
                """
                CREATE TABLE IF NOT EXISTS __snapshots (
                    sequence_num   BIGSERIAL NOT NULL PRIMARY KEY,
                    aggregate_id   VARCHAR(128) NOT NULL UNIQUE,
                    aggregate_type VARCHAR(128) NOT NULL,
                    aggregate_data JSONB NOT NULL DEFAULT '{}'
                );
                """
            )
            await cur.execute("COMMIT;")
            await cur.execute("BEGIN;")
            try:
                await cur.execute("CREATE EXTENSION IF NOT EXISTS plpython3u;")
            except Exception as err:  # pylint: disable=broad-except
                LOG.warning("plpython3u is unavailable: %s", err)
            else:
                await cur.execute(
                    """
                    CREATE OR REPLACE FUNCTION fromisoformat(raw text)
                        RETURNS timestamp with time zone
                    AS $$
                        from datetime import datetime
                        try:
                            return datetime.fromisoformat(raw)
                        except Exception:
                            return None
                    $$ LANGUAGE plpython3u;
                    """
                )
                self.__timestamp_convert = r"fromisoformat({})"
            await cur.execute("COMMIT;")
        except Exception:
            await cur.execute("ROLLBACK;")
            raise

    @asynccontextmanager
    async def get_cursor(self) -> AsyncGenerator[aiopg.Cursor, None]:
        """Get a transaction cursor for the underlying connection."""
        pool = await self.__get_pool()
        async with pool.acquire() as conn:
            async with conn.cursor() as cur:
                yield cur

    @asynccontextmanager
    async def get_transaction(self) -> AsyncGenerator[aiopg.Cursor, None]:
        async with self.get_cursor() as cur:
            await cur.execute("BEGIN;")
            try:
                yield cur
                await cur.execute("COMMIT;")
            except:
                await cur.execute("ROLLBACK;")
                raise

    def __simplify(self, pred: Predicate, type_field: str, data_field: str):
        return cast_simplified_predicate(
            visit_predicate(
                _PostgreSQLSimplifier(
                    type_field, data_field, timestamp_convert=self.__timestamp_convert
                ),
                pred,
            )
        )

    async def load_events(self, query: Predicate = None) -> Iterable[EventBase]:
        """Load events that match the predicate."""
        async with self.get_transaction() as conn:
            events: List[EventBase] = []
            sql_str = "SELECT event_type, event_data from __events"
            params = None
            if query:
                query, where_clause, params = self.__simplify(
                    query, "event_type", "event_data"
                )
                if where_clause is not None:
                    sql_str += f" WHERE {where_clause}"
            sql_str += " ORDER BY sequence_num ASC"
            LOG.info("SQL QUERY: %s", sql_str)
            await conn.execute(sql_str, params)
            async for row in conn:
                evt = EventMeta.construct_named(row[0], row[1])
                if query is None or query(evt):
                    events.append(evt)
            return events

    async def save_events(self, events: Iterable[EventBase]):
        """Save new events."""
        async with self.get_transaction() as conn:
            sql_str = "INSERT INTO __events (event_type, event_data) values (%s, %s);"
            for evt in events:
                params = (
                    evt.__class__.__name__,
                    JSON.dumps(evt.to_dict()),
                )
                await conn.execute(sql_str, params)
            LOG.info("SQL EXEC: %s", sql_str)

    async def save_snapshots(self, snaps: Iterable[AggregateBase]):
        """Save new snapshots."""
        async with self.get_transaction() as conn:
            for snap in snaps:
                snap_typ = type(snap)
                snap_id = f"{snap_typ.__name__}:{getattr(snap, str(snap_typ.__schema__.id_field))}"
                snap_data = JSON.dumps(snap.to_dict())
                sql_str = """
                    INSERT INTO __snapshots (aggregate_id, aggregate_type, aggregate_data)
                        VALUES (%s, %s, %s)
                        ON CONFLICT(aggregate_id) DO UPDATE SET
                            aggregate_data=excluded.aggregate_data;
                    """
                params = (snap_id, snap_typ.__name__, snap_data)
                LOG.info("SQL EXEC: %s", sql_str)
                await conn.execute(sql_str, params)

    async def load_snapshots(self, query: Predicate = None) -> Iterable[AggregateBase]:
        """Load snapshots that match the predicate."""
        async with self.get_transaction() as conn:
            snaps: List[AggregateBase] = []
            sql_str = "SELECT aggregate_type, aggregate_data from __snapshots"
            params = None
            if query:
                query, where_clause, params = self.__simplify(
                    query, "aggregate_type", "aggregate_data"
                )
                if where_clause is not None:
                    sql_str += f" WHERE {where_clause}"
            sql_str += " ORDER BY sequence_num ASC"
            LOG.info("SQL QUERY: %s", sql_str)
            await conn.execute(sql_str, params)
            async for row in conn:
                snap = AggregateMeta.construct_named(row[0], row[1])
                if query is None or query(snap):
                    snaps.append(snap)
            return snaps

    async def close(self):
        """Close underlying connection pool."""
        async with self.__setup.write:
            if self.__pool is not None:
                self.__pool.close()
