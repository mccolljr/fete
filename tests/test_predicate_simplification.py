from flurry.core import predicate as P
from flurry.core.utils import visit_predicate
from flurry.postgres.postgres import _PostgreSQLSimplifier

predicates_to_simplify = {
    "empty_or": P.Or(),
    "empty_is": P.Is(),
    "empty_and": P.And(),
    "empty_where": P.Where(),
    "simple_and": P.And(P.Where(a=1), P.Where(b=2)),
    "simple_or": P.Or(P.Where(a=1), P.Where(b=2)),
    "simple_is": P.Is(str, int, float),
    "simple_where": P.Where(
        a=P.Eq(1),
        b=P.NotEq(2),
        c=P.Less(3),
        d=P.More(4),
        e=P.LessEq(5),
        f=P.MoreEq(6),
        g=P.Between(7, 8),
        h=P.OneOf(9, 10),
    ),
    "complex": P.Or(
        P.Is(int, str, float),
        P.And(
            P.Where(a=P.Eq(1)),
            P.Where(b=P.NotEq(2)),
            P.Where(c=P.Less(3)),
        ),
        P.And(
            P.Where(d=P.More(4)),
            P.Where(e=P.LessEq(5)),
            P.Where(f=P.MoreEq(6)),
        ),
        P.Where(
            g=P.Between(7, 8),
            h=P.OneOf(9, 10),
        ),
        P.And(P.Is(), P.Where()),
    ),
}


# def test_sqlite_simplify(snapshot):
#     visitor = _SqliteSimplifier("type_field", "data_field")
#     for name, pred in predicates_to_simplify.items():
#         result = visit_predicate(visitor, pred)
#         assert result == snapshot(name=name)


def test_postgresql_simplify(snapshot):
    visitor = _PostgreSQLSimplifier("type_field", "data_field")
    for name, pred in predicates_to_simplify.items():
        result = visit_predicate(visitor, pred)
        assert result == snapshot(name=name)
