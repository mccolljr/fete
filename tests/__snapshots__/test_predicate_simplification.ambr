# name: test_postgresql_simplify[complex]
  <class 'tuple'> (
    None,
    '(type_field IN (%s, %s, %s) OR (((data_field ? %s AND data_field->%s = %s::jsonb)) AND (coalesce(data_field->%s <> %s::jsonb, true)) AND ((data_field ? %s AND data_field->%s < %s::jsonb))) OR (((data_field ? %s AND data_field->%s > %s::jsonb)) AND ((data_field ? %s AND data_field->%s <= %s::jsonb)) AND ((data_field ? %s AND data_field->%s >= %s::jsonb))) OR (((data_field ? %s AND data_field->%s >= %s::jsonb) AND (data_field ? %s AND data_field->%s <= %s::jsonb)) AND ((data_field ? %s AND data_field->%s = %s::jsonb) OR (data_field ? %s AND data_field->%s = %s::jsonb))))',
    <class 'list'> [
      'int',
      'str',
      'float',
      'a',
      'a',
      '1',
      'b',
      '2',
      'c',
      'c',
      '3',
      'd',
      'd',
      '4',
      'e',
      'e',
      '5',
      'f',
      'f',
      '6',
      'g',
      'g',
      '7',
      'g',
      'g',
      '8',
      'h',
      'h',
      '9',
      'h',
      'h',
      '10',
    ],
  )
---
# name: test_postgresql_simplify[simple_and]
  <class 'tuple'> (
    None,
    '(((data_field ? %s AND data_field->%s = %s::jsonb)) AND ((data_field ? %s AND data_field->%s = %s::jsonb)))',
    <class 'list'> [
      'a',
      'a',
      '1',
      'b',
      'b',
      '2',
    ],
  )
---
# name: test_postgresql_simplify[simple_is]
  <class 'tuple'> (
    None,
    'type_field IN (%s, %s, %s)',
    <class 'list'> [
      'str',
      'int',
      'float',
    ],
  )
---
# name: test_postgresql_simplify[simple_or]
  <class 'tuple'> (
    None,
    '(((data_field ? %s AND data_field->%s = %s::jsonb)) OR ((data_field ? %s AND data_field->%s = %s::jsonb)))',
    <class 'list'> [
      'a',
      'a',
      '1',
      'b',
      'b',
      '2',
    ],
  )
---
# name: test_postgresql_simplify[simple_where]
  <class 'tuple'> (
    None,
    '((data_field ? %s AND data_field->%s = %s::jsonb) AND coalesce(data_field->%s <> %s::jsonb, true) AND (data_field ? %s AND data_field->%s < %s::jsonb) AND (data_field ? %s AND data_field->%s > %s::jsonb) AND (data_field ? %s AND data_field->%s <= %s::jsonb) AND (data_field ? %s AND data_field->%s >= %s::jsonb) AND ((data_field ? %s AND data_field->%s >= %s::jsonb) AND (data_field ? %s AND data_field->%s <= %s::jsonb)) AND ((data_field ? %s AND data_field->%s = %s::jsonb) OR (data_field ? %s AND data_field->%s = %s::jsonb)))',
    <class 'list'> [
      'a',
      'a',
      '1',
      'b',
      '2',
      'c',
      'c',
      '3',
      'd',
      'd',
      '4',
      'e',
      'e',
      '5',
      'f',
      'f',
      '6',
      'g',
      'g',
      '7',
      'g',
      'g',
      '8',
      'h',
      'h',
      '9',
      'h',
      'h',
      '10',
    ],
  )
---
